Chef 14 compatibility fixes

andrewda opened the pull request:
**Old ChefDK (before)**
- [x] Ensure all ChefSpec tests pass
- [x] Ensure all Inspec tests pass

**New ChefDK**
- [x] Ensure cookstyle is passing
- [x] Apply foodcritic fixes
- [x] Ensure `rake` passes
- [x] Ensure all Inspec tests pass

**Old ChefDK (after)**
- [x] Repeat all tests above using old ChefDK

The new `check-mysql.sh` script verifies connection to MySQL with a timeout (it seems to take 60-240s for MariaDB to start up). This was causing InSpec to occasionally fail.

Commit:
ChefSpec fixes

andrewda commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Chef 14: dobc](https://trello.com/c/jrp9aGkJ/1496-chef-14-dobc)

ramereth commented on files/default/check-mysql.sh:
```diff
@@ -0,0 +1,25 @@
+port=$1
+ip=$2
+
+script="sshpass -p password ssh -p ${port} -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no dobc@localhost -- mysql -h'${ip}' -P'3306' -uroot -p'dobc' -e exit"
```
So the container already has an environment variable which includes the IP address for the host, you can see it by doing this:
```console
$ sshpass -p password ssh -p 33000 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no dobc@localhost export
Warning: Permanently added '[localhost]:33000' (ECDSA) to the list of known hosts.
declare -x DOBC_PASSWORD="password"
declare -x HOME="/home/dobc"
declare -x LANG="en_US.UTF-8"
declare -x LESSOPEN="||/usr/bin/lesspipe.sh %s"
declare -x LOGNAME="dobc"
declare -x MAIL="/var/mail/dobc"
declare -x MYSQL_ENV_GOSU_VERSION="1.10"
declare -x MYSQL_ENV_GPG_KEYS="177F4010FE56CA3336300305F1656F24C74CD1D8"
declare -x MYSQL_ENV_MARIADB_MAJOR="10.4"
declare -x MYSQL_ENV_MARIADB_VERSION="1:10.4.10+maria~bionic"
declare -x MYSQL_ENV_MYSQL_ROOT_PASSWORD="dobc"
declare -x MYSQL_NAME="/dobc-00/mysql"
declare -x MYSQL_PORT="tcp://172.17.0.2:3306"
declare -x MYSQL_PORT_3306_TCP="tcp://172.17.0.2:3306"
declare -x MYSQL_PORT_3306_TCP_ADDR="172.17.0.2"
declare -x MYSQL_PORT_3306_TCP_PORT="3306"
declare -x MYSQL_PORT_3306_TCP_PROTO="tcp"
declare -x OLDPWD
declare -x PATH="/usr/local/bin:/usr/bin"
declare -x PWD="/home/dobc"
declare -x SHELL="/bin/bash"
declare -x SHLVL="1"
declare -x SSH_CLIENT="172.17.0.1 60440 22"
declare -x SSH_CONNECTION="172.17.0.1 60440 172.17.0.3 22"
declare -x USER="dobc"
```
So this should be changed to the following:
```suggestion
script="sshpass -p password ssh -p ${port} -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no dobc@localhost -- mysql -h'${MYSQL_PORT_3306_TCP_ADDR}' -P'3306' -uroot -p'dobc' -e exit"
```

The problem is, the IP you set might be different depending on how things get deployed.

ramereth replied:
So I discovered an issue with this as the environment variable does not get created when you run the command. I tried a variety of things and couldn't get it to show up. I think we should go back to statically setting the IP since it looks like that's what we did before.

andrewda replied:
@ramereth hmm, it seemed to be working with my tests. Are you using:

```bash
sshpass -p password ssh -p ${port} -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no dobc@localhost -- mysql -h'\${MYSQL_PORT_3306_TCP_ADDR}' -P'3306' -uroot -p'dobc' -e exit
```

Note the backslash before `${MYSQL_PORT_3306_TCP_ADDR}`. I think that's crucial, because we want `${MYSQL_PORT_3306_TCP_ADDR}` to be executed where the variable is set (in the container).

ramereth replied:
It keeps failing for me however:
```
  Command: `sshpass -p password ssh -p 33000 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no dobc@localhost
 hostname`
     ×  exit_status is expected to eq 0
     
     expected: 0
          got: 255
     
     (compared using ==)

     ×  stdout is expected to match /^dobc$/
     expected "" to match /^dobc$/
     Diff:
     @@ -1,2 +1,2 @@
     -/^dobc$/
     +""

  Command: `/usr/local/bin/check-mysql.sh 33000`
     ×  exit_status is expected to eq 0
     
     expected: 0
          got: 1
     
     (compared using ==)
```

andrewda replied:
Ah weird, okay. I'll take a look into this

andrewda replied:
Hmm, I'm having a lot of trouble replicating this error. Ran on Chef 13 and 14 three times each, and has passed every time:

```
  Command: `sshpass -p password ssh -p 33000 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no dobc@localhost hostname`
     ✔  exit_status is expected to eq 0
     ✔  stdout is expected to match /^dobc$/
  Command: `/usr/local/bin/check-mysql.sh 33000`
     ✔  exit_status is expected to eq 0
```

Are you seeing this error in 13, 14 or both?

ramereth replied:
This seems to only be happening with 14.

andrewda replied:
Getting a `255` error with SSH seems to suggest that the host is down, so it might be that the container isn't starting up properly (which would explain the failure for both commands). After another 5 or so `kitchen test`s I still haven't been able to duplicate the issue, though...

andrewda replied:
Ah, I just ran it again and got the same error. Looks like it's originating from `start-all.sh` returning error `125`, which might be related to https://github.com/docker/swarm/issues/2835. If you think it'd be best, I can add a retry to `start-all.sh` in case we receive a `125` while starting the containers.

ramereth replied:
Yes, let's try that.

ramereth replied:
This is still failing for me :cry:

andrewda commented:
@ramereth Latest commit adds a retry script, which will retry starting docker containers up to 10 times. I've ran the tests ~8 times with retry and haven't ran into the error yet, so hopefully it increases reliability for you as well.

ramereth committed:
Skip tzinfo init with mysql containers

This undocumented env var [1] seems to fix the issue we've been having with the
mysql container startup taking too long. I don't believe we need this for DOBC
so we can remove this later if we need to.

[1] https://github.com/docker-library/mariadb/blob/master/docker-entrypoint.sh#L196-L201

ramereth reviewed (approved):
Please do a patch bump.

andrewda commented:
!bump patch ~

osuosl-manatee commented:
Jenkins has merged this PR into `master` and has automatically performed a patch-level version bump to v2.0.3.  Have a nice day!
